p2p/src/transport.rs

```diff
@@ -0,0 +1,57 @@
+use std::fmt;
+use std::io::{Read, Write};
+use std::net::SocketAddr;
+
+use tendermint::public_key::PublicKey;
+
+use eyre::Result;
+
+pub struct BindInfo {
+    pub addr: SocketAddr,
+    pub advertise_addrs: Vec<SocketAddr>,
+    pub public_key: PublicKey,
+}
+
+pub struct ConnectInfo {}
+
+#[derive(Clone, Copy, Hash, Eq, PartialEq)]
+pub enum StreamId {
+    Pex,
+}
+
+pub enum Direction<Conn> {
+    Incoming(Conn),
+    Outgoing(Conn),
+}
+
+pub trait Connection: Send {
+    type Error: 'static + fmt::Display + std::error::Error + Send + Sync;
+    type Read: Read;
+    type Write: Write;
+
+    fn advertised_addrs(&self) -> Vec<SocketAddr>;
+    fn close(&self) -> Result<()>;
+    fn local_addr(&self) -> SocketAddr;
+    fn open_bidirectional(
+        &self,
+        stream_id: &StreamId,
+    ) -> Result<(Self::Read, Self::Write), Self::Error>;
+    fn public_key(&self) -> PublicKey;
+    fn remote_addr(&self) -> SocketAddr;
+}
+
+pub trait Endpoint: Send {
+    type Connection;
+
+    fn connect(&self, info: ConnectInfo) -> Result<Self::Connection>;
+    fn listen_addrs(&self) -> SocketAddr;
+}
+
+pub trait Transport {
+    type Connection: Connection;
+    type Endpoint: Endpoint<Connection = <Self as Transport>::Connection>;
+    type Incoming: Iterator<Item = Result<<Self as Transport>::Connection>> + Send;
+
+    fn bind(&self, bind_info: BindInfo) -> Result<(Self::Endpoint, Self::Incoming)>;
```

Comment:
There will definitely be a memory transport. If that will work with  dummy IPs I'm not sure yet. In the end it could be true that the bindinfo is trasnport specific. Will leave it as an open point.